core/src/main/java/org/springframework/security/oauth2/server/authorization/OAuth2AuthorizationService.java

```diff
@@ -42,4 +43,19 @@
 	 */
 	OAuth2Authorization findByTokenAndTokenType(String token, TokenType tokenType);
 
+	/**
+	 * Returns the {@link OAuth2Authorization} containing the provided {@code token},
+	 * or {@code null} if not found.
+	 *
+	 * @param token the token credential
+	 * @return the {@link OAuth2Authorization} if found, otherwise {@code null}
+	 */
+	OAuth2Authorization findByToken(String token);
+
+	/**
+	 * Invalidates the given {@link OAuth2Authorization}
+	 *
+	 * @param authorization {@link OAuth2Authorization} to be invalidated
+	 */
+	void invalidate(OAuth2Authorization authorization);
```

Comment:
I'm not sure we should add this operation here. We will need the ability to revoke access tokens, refresh tokens and possibly authorization codes. This operation doesn't allow us to specify which token to revoke. I wonder if we should have a `revoke` on `OAuth2Authorization` instead. Thoughts?

Reply:
@jgrandja Joe this is the only place where I have a doubt, you mean to say that have a revoke method on the authorization and then we should do
authorization.revoke() and then authorizationService.save(authorization)? Because in that case with our current inmemory implementation one more record will get inserted with revoked status. Instead I thought we can have an invalidate method on the service and let the implementation decide how to handle the revoke. In this default inmemory implementation 
1. we remove the authorization and then
2. add it in revoked status
Other implementations may remove it from DB and handle other revocations etc. 
We can discuss this further Joe